bot/bot.py

```diff
@@ -131,8 +134,26 @@ def clear(self) -> None:
         self._recreate()
         super().clear()
 
+    def _remove_extensions(self) -> None:
+        """Remove all extensions to trigger cog unloads."""
+        extensions = list(self.extensions.keys())
+
+        for ext in extensions:
+            try:
+                self.unload_extension(ext)
+            except Exception:
+                pass
+
     async def close(self) -> None:
         """Close the Discord connection and the aiohttp session, connector, statsd client, and resolver."""
+        # Done before super().close() to allow tasks finish before the HTTP session closes.
+        self._remove_extensions()
+
+        # Wait until all tasks that have to be completed before bot is closing is done
+        log.trace("Waiting for tasks before closing.")
+        await asyncio.gather(*self.closing_tasks)
```

Comment:
While this approach does seem feasible for solving issues when shutting down, I'm concerned about what will happen when cogs are manually reloaded. For example, could there be a race condition between a closing task and some init task?

Reply:
There isn't currently any closing task that really can have dangerous race conditions. This work same way for reloading like this worked before, as this now just add them to list.